### PR TITLE
Organize functionality in categories for the doxygen documentation

### PR DESCRIPTION
**Pull request by gnzlbg**
This commit
- puts the placeholders in the `meta::placeholders` namespace (they are automatically included in the `meta` namespace, but this allows one to easily include the placeholders without including all of `meta`).
- adds more examples to the user guide (partial application, logical operations, lambdas, typical operations on lists, and brief overview of functionality).
- merges back the workarounds into a single file.
- organizes the functionality in the doxygen documentation as follows:

  - Metafunctions
    - Invocation
    - Composition
  - Logical
  - Algorithms
    - Query/search
    - Transformation
  - Datatype
    - List
    - Integral constant
  - Math

Each section of the documentation contains a lazy subcategory with the corresponding metafunctions.

### CODE REVIEW

File: doc/index.md [hunk 1]
```diff
@@ -78,33 +78,115 @@ Note that in the first case we create a metafunction class that will evaluate
...
+
+### Overview
+
+This is a brief overview of the functionality in meta:
+
+- Metafunction: `meta::eval`, `meta::apply`, `meta::defer`, `meta::quote`,
+  `meta::compose`, `meta::bind_front`, `meta::bind_back`, `meta::curry`,
+  `meta::uncurry`, `meta::lambda`.
+- List: `meta::list`, `meta::list_element`,
+  `meta::list_element_c`. `meta::empty`, `meta::size`.
```
Comment (ericniebler) [inline doc/index.md]: meta::front, meta::back

Comment (gnzlbg) [inline doc/index.md]: fixed

**Comment by gnzlbg** [comment]
The complexity of the `find` algorithms is missing. It is O(N) right?

**Comment by ericniebler** [comment]
Yes

**Comment by gnzlbg** [comment]
Added.

File: include/meta/meta.hpp [hunk 2]
```diff
@@ -311,35 +417,47 @@ namespace meta
...
+            };
+        } // namespace detail
+        /// \endcond
 
         /// Applies the Metafunction Class \p C using the types in
         /// the type list \p List as arguments.
+        /// \ingroup invocation
         template <typename C, typename List>
-        using apply_list = eval<apply_list_impl<C, List>>;
+        using apply_list = eval<detail::apply_list_<C, List>>;
```
Comment (ericniebler) [inline include/meta/meta.hpp]: No, I actually intend for `apply_list_impl` to be part of the pulic API, although the name needs work. End users should be able to specialize it to get `as_list`, `apply_list`, `curry`, `uncurry` and all the rest to work with their types. Notice how it's specialized for `integer_sequence`.

File: include/meta/meta.hpp [hunk 3]
```diff
@@ -1577,19 +1843,25 @@ namespace meta
...
+            struct _b;
+            struct _c;
+            struct _d;
+            struct _e;
+            struct _f;
+            struct _g;
+            struct _h;
+            struct _i;
+        } // namespace placeholders
+        using namespace placeholders;
```
Comment (ericniebler) [inline include/meta/meta.hpp]: `inline namespace`

File: readme.md [hunk 4]
```diff
@@ -2,14 +2,15 @@
...
 #include <meta/meta.hpp>
 ```
 
-
-
 For a quick start see Eric Niebler's blog post:
 [A tiny metaprogramming library](http://ericniebler.com/2014/11/13/tiny-metaprogramming-library/). (Note: the names in Meta are different from those describe in the blog post, but the overall design remains the same.)
+
+To generate the up-to-date tutorial and documentation run `make doc` in the
+build directory (requires Doxygen).
```
Comment (ericniebler) [inline readme.md]: ... and LaTeX and dvips and ghostscript.

File: include/meta/meta.hpp (hunk 2 above)
Comment (gnzlbg) [inline include/meta/meta.hpp]: I think boost fusion has an `extension` namespace. What about `extension::apply_list` ?

Comment (ericniebler) [inline include/meta/meta.hpp]: I suppose that's alright.

Comment (ericniebler) [inline include/meta/meta.hpp]: I just took care of this one. Sorry for the merge conflict. :-/

File: include/meta/meta.hpp (hunk 3 above)
Comment (ericniebler) [inline include/meta/meta.hpp]: I took care of this one, too.

**Comment by gnzlbg** [comment]
I somehow managed to screw up the history... now I landed in a git tutorial that says I should never use `git push --force`, while the last couple of them I read when I ran into problems with updating conflicts in pull requests said that the easiest solution was `push --force`. I'm closing this, reforking, and submitting a clean one.

File: readme.md (hunk 4 above)
Comment (gnzlbg) [inline readme.md]: fixed

**Comment by ericniebler** [comment]
Yes, *never* use git push --force. It shouldn't have been needed to deal work a simple merge conflict.

**Comment by ericniebler** [comment]
Also, THANK YOU for doing this work.

**Comment by gabyx** [comment]
This library is so much better than boost::mpl :-) !!!

easier to use!! (also thanks to c++11) and of course the fine work of eric

it would also be handy to have a  

getIdx<T,List> 

function which returns the index for of type T in a List :-)

or is there already such a thing :-), i did it with find<>…

br gabriel






Am 25.02.2015 um 22:42 schrieb Eric Niebler <notifications@github.com>:

**Comment by ericniebler** [comment]
> it would also be handy to have a
> 
> getIdx<T,List>
> 
> function which returns the index for of type T in a List :-)

I've wanted this too. `find` is the way to do this today. Could you open a new issue so I don't forget?

**Comment by pfultz2** [comment]
> I've wanted this too. find is the way to do this today. Could you open a new issue so I don't forget?

The issues aren't open on the this library.

**Comment by ericniebler** [comment]
Oh weird. It's there now.

**Comment by gnzlbg** [comment]
And travis-ci is not set!
And the docs online would be nice too :D

> I've wanted this too. find is the way to do this today.

Pull-request sent, i just called it meta::index<T, List>, but i don't like naming things.

**Comment by pfultz2** [comment]
> Pull-request sent, i just called it meta::index, but i don't like naming things.

Perhaps, `find_index` or `index_of`, but `index` works as well.

**Comment by gnzlbg** [comment]
There is an API decision to be made there. Should it work like `std::find`, and `std::string::find` and return `end` or `npos` (`list::size()` if not found), or should it return `meta::nil_` on not found? 

Both have pros and cons. We should discuss this in the pull-request because it would be an API breaking change.

**Comment by gabyx** [comment]
I would make the names, as concise and short as possible,

even 

meta::list_element    is  a bit long ? 

why not  meta::element

and meta::index  (i like, meta::index_of    has a shitty underscore :-) 

or even shorter  meta::idx   (maybee to short)

for the index:
static_asserts should not be in the library right?  the user should be responsible for this kind…?

i would return list::size()   since it mimics the end iterator for integral_constants,   if it is somekind of “end” type then we are in the iterator world once more =)






Am 25.02.2015 um 23:50 schrieb gnzlbg <notifications@github.com>:

**Comment by ericniebler** [comment]
`list_element` is too long. I would prefer either `get` (like `std::get` for tuples), or `at` (like MPL). Thoughts?

**Comment by pfultz2** [comment]
I would prefer `at`.

**Comment by gnzlbg** [comment]
I prefer at too.

**Pull request closed without merging**